### PR TITLE
HDDS-6967. EC: Handle maintenance replicas in ECUnderReplicationHandler

### PR DESCRIPTION
**Pull request by siddhantsangwan**
## What changes were proposed in this pull request?

Changes to handle under replication when it's because of replicas in maintenance DNs. 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6967

## How was this patch tested?

New unit tests.

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECContainerReplicaCount.java [hunk 1]
```diff
@@ -196,6 +196,29 @@ public Set<Integer> maintenanceIndexes() {
...
+  /**
+   * Gets a set containing all maintenance only indexes. These are replicas
+   * that are only on maintenance nodes, without any copies on healthy nodes.
+   *
+   * @param includePendingAdd true if any indexes that are pending an add
+   * {@link ContainerReplicaOp.PendingOpType#ADD} should be excluded
+   * @return set containing maintenance only indexes or empty set if none are
+   * in maintenance
+   */
+  public Set<Integer> maintenanceOnlyIndexes(boolean includePendingAdd) {
```
Comment (sodonnel) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECContainerReplicaCount.java:208]: I think we should remove the other version of maintenanceOnlyIndexes that returns a list, and make everything use this one. It keeps it consistent with decommissionOnlyIndexes

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestECContainerReplicaCount.java [hunk 2]
```diff
@@ -229,7 +229,7 @@ public void testAdditionalMaintenanceCopiesAllMaintenance() {
             Collections.emptyList(), 1);
     Assertions.assertFalse(rcnt.isSufficientlyReplicated(false));
     Assertions.assertFalse(rcnt.isOverReplicated(true));
-    Assertions.assertEquals(4, rcnt.additionalMaintenanceCopiesNeeded());
+    Assertions.assertEquals(4, rcnt.additionalMaintenanceCopiesNeeded(false));
```
Comment (sodonnel) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestECContainerReplicaCount.java:232]: All the calls I can see in the this test calss call `rcnt.additionalMaintenanceCopiesNeeded(false)` with `false`. We should add at least one test with some pending add and test the `true` condition too.

**Comment by sodonnel** [comment]
The changes look good overall - I just have 2 minor comments. If we can address those it will be good to commit.

**Comment by siddhantsangwan** [comment]
@sodonnel thanks for the review. I've addressed your comments.

**Review by sodonnel — APPROVED** [review]

**Pull request merged**